### PR TITLE
[IAM Policy] CRUD Refinement ,code fixes , UTs and STs

### PR DESCRIPTION
**Pull request by shalakadharap**
# Problem Statement
PR created for refinement changes in policy CRUD code

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 1]
```diff
@@ -64,6 +64,7 @@ public ServerResponse create() {
             policy = policyDAO.find(requestor.getAccount(),
                     requestBody.get("PolicyName"));
         } catch (DataAccessException ex) {
+          LOGGER.error("Failed to create policy");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:67]: Pls log policy name along with error message.
General guideline, always log the message with context of error message with variables and values

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java [hunk 2]
```diff
@@ -116,9 +118,12 @@ public ServerResponse create() {
         policy = policyDAO.find(requestBody.get("PolicyARN"));
       }
       catch (DataAccessException ex) {
+        LOGGER.error("Failed to find the requested policy in ldap");
         return responseGenerator.internalServerError();
       }
-      if (policy == null || !policy.exists()) {
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/controller/PolicyController.java:121]: Can delete policy permission be granted to cross account using IAM Policy, 
Delete Policy is called with credentials and using credentials policy is searched in ldap, so there is no chance that account 1 can see policies of account, in that case why this check is required.

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 1 above)
Comment (bkirunge7) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java:67]: Please add additional info in log message at all places in this PR

File: auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java [hunk 3]
```diff
@@ -29,11 +29,17 @@
       method.invoke(instance, obj);
     }
     catch (ClassNotFoundException e) {
+      LOGGER.error("Failed to save ");
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java:32]: @Override public void save(Map<String, Object> dataMap, Object obj,    --> Object is already part of map 
                             String prefix) throws DataAccessException {
    String methodName = "save"; -->define as const string
    String className = "com.seagates3.dao.ldap." + prefix + "LdapStore";   ----->  Repeated in all methods define as const string
    LOGGER.debug("calling method - " + methodName + " of class - " + className);
    try {
      Class < ? > storeClass = Class.forName(className);
      Object instance = storeClass.newInstance();
      Method method = storeClass.getMethod(methodName, Object.class);
      method.invoke(instance, obj);
    }

File: st/clitests/awsiam_spec.py [hunk 4]
```diff
@@ -62,6 +70,35 @@ def user_tests():
...
+    result = AwsIamTest('Create Policy').create_policy("iampolicy",samplepolicy_testing).execute_test()
+    result.command_response_should_have("iampolicy")
+    arn = (get_arn_from_policy_object(result.status.stdout))
+
+    #create-policy fails if policy with same name already exists
+    AwsIamTest('Create Policy').create_policy("iampolicy",samplepolicy_testing).execute_test(negative_case=True)\
+        .command_should_fail().command_error_should_have("EntityAlreadyExists")
+
+    #get-policy
+    AwsIamTest('Get Policy').get_policy(arn).execute_test().command_is_successful()
```
Comment (bkirunge7) [inline st/clitests/awsiam_spec.py:86]: compare the response of the command, policy name should be part of response

File: st/clitests/awsiam_spec.py [hunk 5]
```diff
@@ -62,6 +70,35 @@ def user_tests():
...
+
+    #list-policies
+    AwsIamTest('List Policies').list_policies().execute_test().command_is_successful()
```
Comment (bkirunge7) [inline st/clitests/awsiam_spec.py:89]: match output response with expected output

File: st/clitests/awsiam_spec.py [hunk 6]
```diff
@@ -62,6 +70,35 @@ def user_tests():
...
+
+    #delete-policy
+    AwsIamTest('Delete Policy').delete_policy(arn).execute_test().command_is_successful()
+
+    #get-policy on non-existing policy
+    AwsIamTest('Get Policy').get_policy(arn).execute_test(negative_case=True).command_should_fail().command_error_should_have("NoSuchEntity")
+
+    #delete-policy on non-existing policy
+    AwsIamTest('Delete Policy').delete_policy(arn).execute_test(negative_case=True).command_should_fail().command_error_should_have("NoSuchEntity")
```
Comment (bkirunge7) [inline st/clitests/awsiam_spec.py:98]: add scenario: delete policy of a different account

File: st/clitests/awsiam_spec.py (hunk 5 above)
Comment (bkirunge7) [reply st/clitests/awsiam_spec.py:89]: match expected number of policy(how many policies listed)

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 2 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java:121]: Before deleting we search policy using ARN not by account. So account1 can try deleting policy of account2
Hence the check
In aws, cross account delete-policy not allowed.

File: st/clitests/awsiam_spec.py (hunk 6 above)
Comment (bkirunge7) [reply st/clitests/awsiam_spec.py:122]: @shalakadharap pls check

File: st/clitests/awsiam_spec.py (hunk 5 above)
Comment (bkirunge7) [reply st/clitests/awsiam_spec.py]: @shalakadharap

File: st/clitests/awsiam_spec.py (hunk 4 above)
Comment (bkirunge7) [reply st/clitests/awsiam_spec.py]: @shalakadharap

File: auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java (hunk 3 above)
Comment (bkirunge7) [reply auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java]: @shalakadharap

File: st/clitests/awsiam_spec.py (hunk 6 above)
Comment (bkirunge7) [reply st/clitests/awsiam_spec.py:122]: > @shalakadharap pls check

@shalakadharap  please check

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 1 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java]: Fixed

File: auth/server/src/main/java/com/seagates3/controller/PolicyController.java (hunk 2 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/controller/PolicyController.java:121]: Fixed new design logic

File: auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java (hunk 3 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java]: Fixed

File: st/clitests/awsiam_spec.py (hunk 4 above)
Comment (shalakadharap) [reply st/clitests/awsiam_spec.py]: Fixed

File: st/clitests/awsiam_spec.py (hunk 5 above)
Comment (shalakadharap) [reply st/clitests/awsiam_spec.py]: Fixed

File: st/clitests/awsiam_spec.py (hunk 6 above)
Comment (shalakadharap) [reply st/clitests/awsiam_spec.py:122]: Added

File: auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java [hunk 7]
```diff
@@ -10,11 +10,11 @@
 interface AuthStore {
 
  public
-  void save(Map<String, Object> dataMap, Object obj,
+  void save(Map<String, Object> dataMap,
             String prefix) throws DataAccessException;
 
  public
-  Object find(String keyToFind, Object obj,
+  Object find(String keyToFind, Object obj, Object obj2,
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java:17]: what is obj2, can you name it based on context

File: auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java [hunk 8]
```diff
@@ -15,99 +15,124 @@
...
     }
-    return null;
   }
 
   @Override public List findAll(String strToFind, Object obj,
                                 String prefix) throws DataAccessException {
-    String methodName = "findAll";
-    String className = "com.seagates3.dao.ldap." + prefix + "LdapStore";
-    LOGGER.debug("calling method - " + methodName + " of class - " + className);
+    String className = CLASS_PACKAGE + prefix + CLASS_SUFFIX;
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java:89]: className can be const and defined at one place

File: auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java [hunk 9]
```diff
@@ -74,8 +74,9 @@ void save(Object obj) throws DataAccessException {
   }
 
  public
-  Policy find(Object obj) throws DataAccessException {
+  Policy find(Object obj, Object obj2) throws DataAccessException {
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java:77]: rename obj2 with context

File: auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java (hunk 8 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/dao/ldap/LdapStore.java:89]: prefix is coming as parameter here.
So declared it in every method

File: auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java (hunk 7 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java:17]: obj2 can be any object here. So cant rename with specific context here

File: auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java (hunk 9 above)
Comment (bkirunge7) [reply auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java:77]: Policy find(Object policyarn, Object account)

Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/dao/ldap/PolicyLdapStore.java]: Fixed

File: auth/server/src/main/java/com/seagates3/dao/ldap/FileStore.java [hunk 10]
```diff
@@ -54,11 +55,12 @@
...
-      LOGGER.error("Exception occurred while saving policy into file - " + e);
+      LOGGER.error("Exception occurred while saving data into file - " + e);
+      throw new DataAccessException("Exception occurred while saving file");
     }
   }
 
-  @Override public Object find(String key, Object obj,
+  @Override public Object find(String key, Object obj, Object obj2,
                                String prefix) throws DataAccessException {
     return savedDataMap.get(key);
```
Comment (bkirunge7) [inline auth/server/src/main/java/com/seagates3/dao/ldap/FileStore.java:65]: obj and obj2 are not used, are they required here

File: auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java (hunk 7 above)
Comment (bkirunge7) [reply auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java:17]: can you mention example values passed as obj and obj2

File: auth/server/src/main/java/com/seagates3/dao/ldap/FileStore.java (hunk 10 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/dao/ldap/FileStore.java:65]: No. Not required but method prototype made generic for all AUthStorages in AUthStore class

File: auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java (hunk 7 above)
Comment (shalakadharap) [reply auth/server/src/main/java/com/seagates3/dao/ldap/AuthStore.java:17]: We can get this if we perform (ctrl + T) on method.

**Review by bkirunge7 — APPROVED** [review]

**Review by shrimetta — APPROVED** [review]

**Review by rohitkolapkar — APPROVED** [review]

**Pull request merged**